### PR TITLE
Reimplementing external field

### PR DESCRIPTION
**Pull request by ilfreddy**
The external potential is reintroduced as a simple additional operator in the Fock Operator.

- [x] New input section
- [x] Add external potential to `FockOperator`
- [x] Test

### CODE REVIEW

File: src/qmoperators/one_electron/ElectricFieldOperator.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * moment. The operator is simply implemented as a scalar product of
+ * the dipole moment operator with the electric field vector.
+ *
+ */
+
+namespace mrchem {
+
+class ElectricFieldOperator final : public RankZeroTensorOperator {
+public:
+ ElectricFieldOperator(Eigen::Vector3D f)
```
Comment (stigrj) [inline src/qmoperators/one_electron/ElectricFieldOperator.h]: This argument should be a `(const Eigen::Vector3d &f)`. We should probably make such a typedef for a 3D coordinate to use in generar, currently we are using `const double *r`, which is a bit primitive...

File: src/qmoperators/two_electron/FockOperator.cpp [hunk 2]
```diff
@@ -139,13 +157,15 @@ SCFEnergy FockOperator::trace(OrbitalVector &Phi, const ComplexMatrix &F) {
...
     if (this->xc   != 0) E_xc  =  this->xc->getEnergy();
     if (this->xc   != 0) E_xc2 =  this->xc->trace(Phi).real();
+    if (this->ext  != 0) E_ext =  this->ext->trace(Phi).real();
 
     double E_eex    = E_ee  + E_x;
     double E_orbxc2 = E_orb - E_xc2;
-    E_kin = E_orbxc2 - 2.0*E_eex - E_en;
-    E_el  = E_orbxc2 -     E_eex + E_xc;
+    E_kin = E_orbxc2 - 2.0*E_eex - E_en - E_ext;
+    E_el  = E_orbxc2 -     E_eex + E_xc + E_ext;
```
Comment (stigrj) [inline src/qmoperators/two_electron/FockOperator.cpp]: Should `E_ext` be added here? It should already be present in `E_orb`

Comment (ilfreddy) [reply src/qmoperators/two_electron/FockOperator.cpp]: No it should not. I have removed it after that commit, but still it did not help. The result changes but it does not seem correct anyway :-(

Comment (stigrj) [reply src/qmoperators/two_electron/FockOperator.cpp]: Hmm, how far off?

Comment (ilfreddy) [reply src/qmoperators/two_electron/FockOperator.cpp]: Too far. Dipole: 1.11au. Finite Difference (E(+)-E(-))/2h: 0.84au
And it is origin dependent. This are the values in the centre of charge such that the nuclear contribution is zero.

File: src/qmoperators/two_electron/FockOperator.cpp [hunk 3]
```diff
@@ -99,18 +109,21 @@ void FockOperator::rotate(const ComplexMatrix &U) {
...
+    double E_xc2 = 0.0; // Trace of the XC operator
+    double E_ext = 0.0; // External field contribution to the electronic energy
+    double E_nex = 0.0; // External field contribution to the nuclear energy
 
     // Nuclear part
     if (this->nuc != 0) {
+        std::cout << "Computing nuc conts " << ext << std::endl;
         Nuclei &nucs = this->nuc->getNuclei();
         int nNucs = nucs.size();
         for (int i = 0; i < nNucs; i++) {
```
Comment (stigrj) [inline src/qmoperators/two_electron/FockOperator.cpp]: Can you try to move this part into a `trace(nucs)` function in `NuclearOperator`?

Comment (ilfreddy) [reply src/qmoperators/two_electron/FockOperator.cpp]: If you mean the comment, that as just a debugging thing and can be removed. If you mean the lines under the `if`statement, I will have a look.

Comment (stigrj) [reply src/qmoperators/two_electron/FockOperator.cpp]: The latter :slightly_smiling_face:

File: src/qmoperators/one_electron/NuclearOperator.cpp [hunk 4]
```diff
@@ -59,4 +59,28 @@ void NuclearPotential::clear() {
     clearApplyPrec(); // apply_prec = -1
 }
 
+    /** @brief computes the repulsion self energy of a set of nuclei
+     * 
+     * @param[in] nucs the set of nuclei
+     *
+     */
+    
+double NuclearOperator::trace(const Nuclei &nucs) {
```
Comment (stigrj) [inline src/qmoperators/one_electron/NuclearOperator.cpp]: This function should rather compute the repulsion between `*this` nuclear potential and the provided charge distribution (in this case a collection of point charges), e.i. `nuc_i` should be taken from `this->nuclei` and `nuc_j` from the `nucs` input. However, then the second loop should _not_ be truncated, which results in double counting including self-repulsion whenever `this->nuclei` coincides with the input `nucs`. How do we make this generally applicable?

File: src/qmoperators/one_electron/NuclearOperator.cpp [hunk 5]
```diff
@@ -59,4 +60,27 @@ void NuclearPotential::clear() {
     clearApplyPrec(); // apply_prec = -1
 }
 
+/** @brief computes the repulsion self energy of a set of nuclei
+ * 
+ * @param[in] nucs the set of nuclei
+ *
+ */    
+double NuclearOperator::trace(const Nuclei &nucs) {
```
Comment (stigrj) [inline src/qmoperators/one_electron/NuclearOperator.cpp]: This function should rather compute the repulsion between `*this` nuclear potential and the provided charge distribution (in this case a collection of point charges), e.i. `nuc_i` should be taken from `this->nuclei` and `nuc_j` from the `nucs` input. However, then the second loop should _not_ be truncated, which results in double counting including self-repulsion whenever `this->nuclei` coincides with the input `nucs`. How do we make this generally applicable?

Comment (ilfreddy) [reply src/qmoperators/one_electron/NuclearOperator.cpp]: I think there should then be two functions. One for the self-repulsion interaction (not sure it should be part of the NuclearOperator either... and one for the interaction of NuclearOperator with a second set of charge (no double-counting here).

**Comment by ilfreddy** [comment]
@stigrj do you like it better with a chem_utils file which computes the nuclear self-repulsion?

**Comment by stigrj** [comment]
You need to rebase and resolve some conflicts before we can proceed

**Comment by ilfreddy** [comment]
I'm on it now...

**Comment by ilfreddy** [comment]
This can be merged in now.

File: src/SCFDriver.cpp [hunk 6]
```diff
@@ -352,42 +375,58 @@ void SCFDriver::setup() {
...
-        xcfun->setDensityCutoff(dft_cutoff);
+        fock->setExchangeOperator(K);
+    }
+    //For hybrid DFT we need a partial HF exchange
+    if (wf_method == "DFT" and (dft_x_fac > mrcpp::MachineZero)) {
+        K = new ExchangeOperator(*P, *phi, dft_x_fac);
+        fock->setExchangeOperator(K);
+    }
+    //For DFT we need the XC operator
+    if (wf_method == "DFT") {
```
Comment (stigrj) [inline src/SCFDriver.cpp]: Please clean up this section (remove `der_dft`, print statement and comment). Also, the XCFunctional is not set up correctly, you need the `setUseGamma()` `setDensityCutoff()` `evalSetup(1)` and `setupInitialGrid()`.

Comment (ilfreddy) [reply src/SCFDriver.cpp]: OK, but why did you remove those from the constructor?

Comment (stigrj) [reply src/SCFDriver.cpp]: Yes, `gamma` and `cutoff` could have been in the constructor, but we anyway have to actively set the functionals after construction, so I decided to include only the necessary (`const`) parameters in the constructor arguments

**Comment by ilfreddy** [comment]
Fixed and checked that He (BLYP) and Li (PBE) give reasonable results.

**Pull request merged**